device/core/src/sak_authentication_provider.ts

```diff
@@ -0,0 +1,130 @@
+  type: AuthenticationType = AuthenticationType.Token;
+  automaticRenewal: boolean = true;
+
+  private _tokenValidTimeInSeconds: number = 3600;   // 1 hour
+  private _tokenRenewalMarginInSeconds: number = 900; // 15 minutes
+
+  private _currentTokenExpiryTimeInSeconds: number;
+  private _renewalTimeout: NodeJS.Timer;
+
+  private _credentials: DeviceCredentials;
+
+  /**
+   * @private
+   *
+   * Initializes a new instance of the TokenAuthenticationProvider - users should only use the factory methods though.
+   *
+   * @param credentials                    Credentials to be used by the device to connect to the IoT hub.
+   * @param tokenValidTimeInSeconds        [optional] The number of seconds for which a token is supposed to be valid.
+   * @param tokenRenewalMarginInSeconds    [optional] The number of seconds before the end of the validity period during which the `SharedAccessKeyAuthenticationProvider` should renew the token.
+   */
+  constructor(credentials: DeviceCredentials, tokenValidTimeInSeconds?: number, tokenRenewalMarginInSeconds?: number) {
+    super();
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_001: [The `constructor` shall create the initial token value using the `credentials` parameter.]*/
+    this._credentials  = credentials;
+
+    if (tokenValidTimeInSeconds) this._tokenValidTimeInSeconds = tokenValidTimeInSeconds;
+    if (tokenRenewalMarginInSeconds) this._tokenRenewalMarginInSeconds = tokenRenewalMarginInSeconds;
+
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_002: [The `constructor` shall start a timer that will automatically renew the token every (`tokenValidTimeInSeconds` - `tokenRenewalMarginInSeconds`) seconds if specified, or 45 minutes by default.]*/
+    this._renewToken();
+  }
+
+  /**
+   * Gets the current device credentials that are valid and were generated from the connectiom string.
+   *
+   * @param callback function that will be called with either an error or a set of device credentials that can be used to authenticate with the IoT hub.
+   */
+  getDeviceCredentials(callback: (err: Error, credentials: DeviceCredentials) => void): void {
+    if (this._shouldRenewToken()) {
+      this._renewToken();
+    }
+
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_003: [The `getDeviceCredentials` should call its callback with a `null` first parameter and a `DeviceCredentials` object as a second parameter, containing the latest valid token it generated.]*/
+    callback(null, this._credentials);
+  }
+
+  /**
+   * Sets the current security token for the device. Note that it will be overwritten by the timer that is running in the background.
+   * If you're looking for an AuthenticationProvider where you maintain the lifecycle of the token yourself, you should look at the `SharedAccessSignatureAuthorizationProvider` class.
+   *
+   * @param sharedAccessSignature shared access signature that should be stored.
+   */
+  updateSharedAccessSignature(sharedAccessSignature: string): void {
```

Comment:
>updateSharedAccessSignature [](start = 2, length = 27)

suggest renaming this function to shootYourselfInTheFoot().  Or maybe it should be removed.  If an interface definition requires it to be present, would it be better to throw?

Reply:
maybe. technically it could work though, I'm wary of throwing in cases where we can just make it work

---
In reply to: [157021746](https://github.com/Azure/azure-iot-sdk-node/pull/179#discussion_r157021746) [](ancestors = 157021746)

Reply:
this is cosmetic, but I don't see the value here.  If using a SAS provider will work better, why let the user do this with an SAK provider, especially if it's not going to work well?

---
In reply to: [157030877](https://github.com/Azure/azure-iot-sdk-node/pull/179#discussion_r157030877) [](ancestors = 157030877,157021746)

Reply:
removed 

---
In reply to: [157041863](https://github.com/Azure/azure-iot-sdk-node/pull/179#discussion_r157041863) [](ancestors = 157041863,157030877,157021746)